servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/LoadBalancerBuilder.java

```diff
@@ -0,0 +1,123 @@
+
+    // TODO: these healthCheck* methods should be moved into their own OutlierDetection configuration instance
+    //  and much like the LoadBalancingPolicy, we should be able to add `OutlierDetectionPolicy`s
+    /**
+     * Configure an interval for health checking a host that failed to open connections. If no interval is provided
+     * using this method, a default value will be used.
+     * <p>
+     * {@link #healthCheckFailedConnectionsThreshold(int)} can be used to disable the health checking mechanism
+     * and always consider all hosts for establishing new connections.
+     *
+     * @param interval interval at which a background health check will be scheduled.
+     * @param jitter the amount of jitter to apply to each retry {@code interval}.
+     * @return {@code this}.
+     * @see #healthCheckFailedConnectionsThreshold(int)
+     */
+    LoadBalancerBuilder<ResolvedAddress, C> healthCheckInterval(Duration interval, Duration jitter);
+
+    /**
+     * Configure an interval for re-subscribing to the original events stream in case all existing hosts become
+     * unhealthy.
+     * <p>
+     * In situations when there is a latency between {@link ServiceDiscoverer} propagating the updated state and all
+     * known hosts become unhealthy, which could happen due to intermediate caching layers, re-subscribe to the
+     * events stream can help to exit from a dead state.
+     * <p>
+     * {@link #healthCheckFailedConnectionsThreshold(int)} can be used to disable the health checking mechanism
+     * and always consider all hosts for establishing new connections.
+     *
+     * @param interval interval at which re-subscribes will be scheduled.
+     * @param jitter the amount of jitter to apply to each re-subscribe {@code interval}.
+     * @return {@code this}.
+     * @see #healthCheckFailedConnectionsThreshold(int)
+     */
+    LoadBalancerBuilder<ResolvedAddress, C> healthCheckResubscribeInterval(Duration interval, Duration jitter);
+
+    /**
+     * Configure a threshold for consecutive connection failures to a host. When the {@link LoadBalancer}
+     * consecutively fails to open connections in the amount greater or equal to the specified value,
+     * the host will be marked as unhealthy and connection establishment will take place in the background
+     * repeatedly until a connection is established. During that time, the host will not take part in
+     * load balancing selection.
+     * <p>
+     * Use a negative value of the argument to disable health checking.
+     *
+     * @param threshold number of consecutive connection failures to consider a host unhealthy and eligible for
+     * background health checking. Use negative value to disable the health checking mechanism.
+     * @return {@code this}.
+     * @see #backgroundExecutor(Executor)
+     */
+    LoadBalancerBuilder<ResolvedAddress, C> healthCheckFailedConnectionsThreshold(int threshold);
```

Comment:
I agree and think we can take it a bit further as well. I envisioned something that parallels the `LoadBalancingPolicy` in that it will be generic and really a factory of an outlier detector but there can also be multiple of them. There is a TODO comment on the first health-check related method describing that.

Reply:
Nice, I missed that comment :)